crates/mun_compiler/src/diagnostics.rs

```diff
@@ -3,111 +3,510 @@ use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
+                    highliht_range.1 - first_line_offset,
+                ),
+                text_part,
+            ),
+    )
+}
 
 /// Constructs diagnostic messages for the given file.
-pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Diagnostic> {
+pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Snippet> {
     let parse = db.parse(file_id);
+
     let mut result = Vec::new();
 
-    result.extend(parse.errors().iter().map(|err| Diagnostic {
-        level: Level::Error,
-        loc: err.location(),
-        message: format!("Syntax Error: {}", err),
+    let mut source_code = db.file_text(file_id).to_string();
+    replace_tabs(&mut source_code);
+    let source_code_len = source_code.len();
+
+    let relative_file_path = db.file_relative_path(file_id).display().to_string();
+
+    let line_index = db.line_index(file_id);
+
+    result.extend(parse.errors().iter().map(|err| {
+        let first_line = line_index.line_col(err.location().offset()).line;
+        let first_line_offset = line_index.line_offset(first_line);
+
+        let text_part = line_index
+            .text_part(
+                first_line,
+                line_index.line_col(err.location().end_offset()).line,
+                &source_code,
+                source_code_len,
+            )
+            .unwrap();
+
+        Snippet {
+            title: Some(Annotation {
+                label: Some("syntax error".to_string()),
+                id: None,
+                annotation_type: AnnotationType::Error,
+            }),
+            footer: vec![],
+            slices: vec![Slice {
+                source: text_part.to_string(),
+                line_start: first_line as usize + 1,
+                origin: Some(relative_file_path.clone()),
+                fold: true,
+                annotations: vec![SourceAnnotation {
+                    label: format!("syntax error: {}", err),
+                    annotation_type: AnnotationType::Error,
+                    range: {
+                        let mut range = construct_range(
+                            (
+                                err.location().offset().to_usize(),
+                                err.location().end_offset().to_usize(),
+                            ),
+                            text_part,
+                            first_line_offset,
+                        );
+                        // adds here one to make highlight range visible in output
+                        range.1 += 1;
+                        range
+                    },
+                }],
+            }],
+        }
     }));
 
     let result = RefCell::new(result);
     let mut sink = DiagnosticSink::new(|d| {
-        result.borrow_mut().push(Diagnostic {
-            level: Level::Error,
-            loc: d.highlight_range().into(),
-            message: d.message(),
+        let first_line = line_index.line_col(d.highlight_range().start()).line;
```

Comment:
Definitly create more helper functions here, there is a LOT of code duplication in the creation of the snippets.